Can't have `load_best_model_at_end` with explicit `test_datasets`

### Please check that this issue hasn't been reported before.

- [X] I searched previous [Bug Reports](https://github.com/OpenAccess-AI-Collective/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

Ideally, it should be possible to have `load_best_model_at_end: True` with `test_datasets: [...]` as it is already allowed with `val_set_size`. 

Not sure why it was restricted, tried doing a git blame, but there was a massive refactor so hard to trace

### Current behaviour

Currently it crashes with
```
.../lib/python3.10/site-packages/transformers/trainer_callback.py", line 578, in on_train_begin
    assert args.load_best_model_at_end, "EarlyStoppingCallback requires load_best_model_at_end = True"
AssertionError: EarlyStoppingCallback requires load_best_model_at_end = True
```

because load_best_model_at_end gets disabled

https://github.com/OpenAccess-AI-Collective/axolotl/blob/ea00dd08528399405943f880b27ecc60e4d83cdf/src/axolotl/core/trainer_builder.py#L849-L859

### Steps to reproduce

Train qlora model but set 

```
test_datasets:
 - ...
val_set_size: 0
load_best_model_at_end: True
```

### Config yaml

_No response_

### Possible solution

Slightly modify the condition to allow load_best_model_at_end when either test_datasets is set or val_set_size > 0

### Which Operating Systems are you using?

- [X] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.10

### axolotl branch-commit

main/ea00dd08528399405943f880b27ecc60e4d83cdf

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this bug has not been reported yet.
- [X] I am using the latest version of axolotl.
- [X] I have provided enough information for the maintainers to reproduce and diagnose the issue.

--- a/src/axolotl/core/trainer_builder.py
+++ b/src/axolotl/core/trainer_builder.py
@@ -883,16 +883,10 @@ class HFCausalTrainerBuilder(TrainerBuilderBase):
                 self.cfg.load_best_model_at_end is not False
                 or self.cfg.early_stopping_patience
             )
-            and
-                (
-                    (
-                        not self.cfg.test_datasets and self.cfg.val_set_size > 0
-                    )
-                    or
-                    (
-                        self.cfg.test_datasets and self.cfg.val_set_size == 0
-                    )
-                )
+            and (
+                (not self.cfg.test_datasets and self.cfg.val_set_size > 0)
+                or (self.cfg.test_datasets and self.cfg.val_set_size == 0)
+            )
             and self.cfg.save_steps
             and self.cfg.eval_steps
             and self.cfg.save_steps % self.cfg.eval_steps == 0

--- a/src/axolotl/core/trainer_builder.py
+++ b/src/axolotl/core/trainer_builder.py
@@ -883,8 +883,16 @@ class HFCausalTrainerBuilder(TrainerBuilderBase):
                 self.cfg.load_best_model_at_end is not False
                 or self.cfg.early_stopping_patience
             )
-            and not self.cfg.test_datasets
-            and self.cfg.val_set_size > 0
+            and
+                (
+                    (
+                        not self.cfg.test_datasets and self.cfg.val_set_size > 0
+                    )
+                    or
+                    (
+                        self.cfg.test_datasets and self.cfg.val_set_size == 0
+                    )
+                )
             and self.cfg.save_steps
             and self.cfg.eval_steps
             and self.cfg.save_steps % self.cfg.eval_steps == 0

